url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/cmdlib/BeanLibrary.java

```diff
@@ -0,0 +1,560 @@
+
+                            Object lVal = lGetter.invoke(aBase, lArgs);
+                            return resolveCanonical(aPieces, aCurrPiece + 1, lVal);
+                        }
+                    }
+                }
+                catch(CommandException e)
+                {
+                    throw e;
+                }
+                catch (Exception e)
+                {
+                    throw new CommandException("ERROR - calling getter method.", e);
+                }
+            }
+
+            // Oops, no property found.
+            // What should we try next ?
+            // TODO: alternative lookups.
+            // - based on annotations.
+            // - ...
+            throw new CommandException("ERROR - no property.");
+        }
+        else if(aBase instanceof List && lIdx > 0)
+        {
+            List lList = (List) aBase;
+            if(lIdx > lList.size() - 1)
+                throw new CommandException("ERROR index out of bounds.");
+            return resolveCanonical(aPieces, aCurrPiece + 1, lList.get(lIdx));
+        }
+        else throw new CommandException("ERROR - null.");
+    }
+
+    private static List<String> canonize(String aPath)
+            throws CommandException
+    {
+        final String[] lPieces = aPath.split("/");
+        final List<String> lResult= new ArrayList<String>(lPieces.length);
+        for(String lPiece : lPieces)
+        {
+            if(".".equals(lPiece))
+                // Just skip this.
+                continue;
+            else if("..".equals(lPiece))
+            {
+                // Delete the last part and go back to the previous part.
+                int lResultSize = lResult.size();
+                if(lResultSize > 0) lResult.remove(lResultSize - 1);
+            }
+            else
+            {
+                // Note: Could be done more efficiently using a custom parser.
+                final Matcher m = PAT_FIELDNAME_BIS.matcher(lPiece);
+                if(!m.matches())
+                    throw new CommandException(String.format("Illegal path '%s'.", lPiece));
+                // Fetch the prefix.
+                final String lName = m.group(1);
+                if(lName != null) lResult.add(lName);
+                // Get the indices.
+                int lStart = m.end(1)<0?0:m.end(1);
+                final String[] lIndices = lPiece.substring(lStart).split("\\[");
+                // Skip the first empty string.
+                for(int i = 1; i < lIndices.length; i++)
+                {
+                    String lIdx = lIndices[i];
+                    int lClosing = lIdx.lastIndexOf(']');
+                    if(lClosing > 0) lIdx = lIdx.substring(0, lClosing).trim();
+                    lResult.add(lIdx);
+                }
+            }
+        }
+        return lResult;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void update(Object aTarget, Object aProp, Object aVal)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.